core/store/src/cold_storage.rs

```diff
@@ -0,0 +1,243 @@
+use crate::columns::DBKeyType;
+use crate::{DBCol, Store, StoreUpdate};
+
+use borsh::BorshDeserialize;
+use near_primitives::types::BlockHeight;
+use std::collections::HashMap;
+use std::io;
+use strum::IntoEnumIterator;
+
+type StoreKey = Vec<u8>;
+type StoreValue = Option<Vec<u8>>;
+type StoreCache = HashMap<(DBCol, StoreKey), StoreValue>;
+
+struct StoreWithCache<'a> {
+    store: &'a Store,
+    cache: StoreCache,
+}
+
+impl StoreUpdate {
+    fn copy_from_store(
+        &mut self,
+        store: &mut StoreWithCache,
+        col: DBCol,
+        key: StoreKey,
+    ) -> io::Result<()> {
+        let data = store.get(col, &key)?;
+        if let Some(value) = data {
+            if col.is_insert_only() {
+                self.insert(col, &key, &value);
+            } else if col.is_rc() {
+                self.increment_refcount(col, &key, &value);
+            } else {
+                self.set(col, &key, &value);
+            }
```

Comment:
We probably can just use `insert` for all of those cases.  Just comment that in cold storage reference counts are stripped.

Actually I’ve been even wondering if extending StoreUpdate is the best approach here.  StoreUpdate doesn’t really give us any features that we need and if we go below the abstraction layer straight to ColdDatabase we might end up simplifying bunch of code.  Rather than having to go through ColdDatabase::write we could have ColdDatabase::raw_write which just takes `Vec<Box<[u8]>, Box<[u8]>>` and does raw writes to the database.

One potential downside is that adjusting keys would need to happen here.  That’s not a problem in and of itself but currently it’s nicely isolated in colddb.rs.

Reply:
Don't want to include raw_write in this PR.
But yeah, StoreUpdate actually makes things more difficult because or is_insert_only and is_rc checks. Changed this part to creating a DBTransaction for every column and writing it to cold database.

Reply:
Actually need special case for rc columns after all. I looked at using insert, but thought it wouldn't work. But I may be wrong.
Closing this conversation, as code changed significantly. Look at copy_from_store for the new version of this.